File: fsharp/clock/Clock.fs
module Clock

type Clock = {h : int; m : int}

let rec create hours minutes = 
    match (hours < 0, minutes < 0) with
    | (false, false)    ->
            {h = (hours + (minutes / 60)) % 24
             m = minutes % 60}
    | (false, true)     ->
            create (hours - 1) (minutes + 60)
    | (true, false)     ->
            create (hours  + 24) minutes
    | (true, true)     ->
            create (hours + 23) (minutes + 60)

let add minutes clock = 
    create clock.h (clock.m + minutes)

let subtract minutes clock = 
    create clock.h (clock.m - minutes)

let display clock = 
    // interpolated strings need F# 5.0
    $"%02d{clock.h}:%02d{clock.m}"

File: r/clock/1/clock.R
new_clock <- function(hours, minutes) {
  clean_clock(hours * 60L + minutes)
}

clean_clock <- function(mins_since_midnight) {
  mins_in_day <- 24L * 60L
  while (mins_since_midnight < 0) {
    mins_since_midnight = mins_since_midnight + mins_in_day
  }
  mins_since_midnight %% mins_in_day
}

display <- function(clock) {
  sprintf("%02d:%02d", clock %/% 60L, clock %% 60L)
}

add <- function(clock, minutes) {
  clean_clock(clock + minutes)
}

subtract <- function(clock, minutes) {
  add(clock, -minutes)
}
